KSGraphVertex.java

```diff
@@ -0,0 +1,269 @@
+                    }
+                }
+                else
+                {
+                    newGraphVertex = (KSGraphVertexWeighted) edge.GetTargetVertex();
+                    KSGraphVertexBackTrack backTrack = new KSGraphVertexBackTrack(newGraphVertex);
+                    backTrack.GetBackTrackList().addAll(queuedVertexes.get(currentVertex.GetVertexName()).GetBackTrackList());
+                    backTrack.GetBackTrackList().add(currentVertex);
+                    queuedVertexes.put(newGraphVertex.GetVertexName(), backTrack);
+                    newGraphVertex.SetVertexWeight(currentVertex.GetVertexWeight()+1);
+                }
+
+                vertexQueue.add((KSGraphVertexWeighted) edge.GetTargetVertex());
+            }
+        }
+
+        //
+        // CodeGuru should have ensured that we are not de-referencing a null pointer
+        //
+        return queuedVertexes.get(targetVertex.GetVertexName()).GetBackTrackList();
+        /*ArrayList<KSGraphVertex> path = null;
+        KSGraphVertexBackTrack queryTarget = queuedVertexes.get(targetVertex.GetVertexName());
+        if(queryTarget!=null)
+        {
+            path = queryTarget.GetBackTrackList();
+        }
+        return path;*/
+    }
+
+    public  HashMap<String, KSGraphVertex> CloneGraph()
+    {
+        HashMap<String, KSGraphVertex> clonedGraph = new HashMap<String, KSGraphVertex>();
+        Queue<ArrayList<KSGraphVertex>> graphQueue = new LinkedList<ArrayList<KSGraphVertex>>();
+        ArrayList<KSGraphVertex> pairGraphVertex = new ArrayList<KSGraphVertex>();
+        KSGraphVertex newGraphVertex = new KSGraphVertex(this.GetVertexName());
+
+        clonedGraph.put(newGraphVertex.GetVertexName(), newGraphVertex);
+        pairGraphVertex.add(this);
+        pairGraphVertex.add(newGraphVertex);
+        graphQueue.add(pairGraphVertex);
+
+        while(!graphQueue.isEmpty())
+        {
+            pairGraphVertex = graphQueue.remove();
+            KSGraphVertex originalVertex = pairGraphVertex.get(0);
+            KSGraphVertex clonedVertex = pairGraphVertex.get(1);
+            for(KSGraphEdge edge : originalVertex.GetOutgoingEdges())
+            {
+                if(clonedGraph.containsKey(edge.GetTargetVertex().GetVertexName()))
+                {
+                    newGraphVertex = clonedGraph.get(edge.GetTargetVertex().GetVertexName());
+                }
+                else
+                {
+                    newGraphVertex = new KSGraphVertex(edge.GetTargetVertex().GetVertexName());
+                    pairGraphVertex = new ArrayList<KSGraphVertex>();
+                    pairGraphVertex.add(edge.GetTargetVertex());
+                    pairGraphVertex.add(newGraphVertex);
+                    graphQueue.add(pairGraphVertex);
+                    clonedGraph.put(newGraphVertex.GetVertexName(), newGraphVertex);
+                }
+
+                clonedVertex.ConnectToTarget(newGraphVertex);
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 165:188, 207:230.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.